### PR TITLE
feat: blocklog pruning

### PR DESCRIPTION
**Pull request by dessaya**

### CODE REVIEW

File: packages/evm/jsonrpc/evmchain.go [hunk 1]
```diff
@@ -573,13 +576,15 @@ func blockchainDB(chainState state.State) *emulator.BlockchainDB {
...
 	gasFeePolicy := governance.MustGetGasFeePolicy(govPartition)
+	blockKeepAmount := governance.GetBlockKeepAmount(govPartition)
 	bdbPartition := subrealm.NewReadOnly(
 		chainState,
 		kv.Key(evm.Contract.Hname().Bytes())+evm.KeyEVMState+emulator.KeyBlockchainDB,
 	)
 	return emulator.NewBlockchainDB(
 		buffered.NewBufferedKVStore(bdbPartition),
 		gas.EVMBlockGasLimit(gasLimits, &gasFeePolicy.EVMGasRatio),
+		int(blockKeepAmount),
```
Comment (jorgemmsilva) [inline packages/evm/jsonrpc/evmchain.go]: why the mismatch types for `blockKeepAmount`? `int32` in gov contract and `int` in blockchaindb?

File: documentation/docs/guide/evm/using.md [hunk 2]
```diff
@@ -22,8 +22,7 @@ keywords:
 ## 1. Deploy an IOTA Smart Contracts Chain
 
 When [deploying an IOTA Smart Contracts chain](../chains_and_nodes/setting-up-a-chain.md), EVM support is automatically
-added with the default configuration. The `wasp-cli chain deploy` command accepts some EVM-specific options listed
-below:
+added with the default configuration. The `wasp-cli chain deploy` command accepts the following EVM-specific option:
 
 * `--evm-chainid <n>`: EVM chain ID (default: 1074).
```
Comment (jorgemmsilva) [inline documentation/docs/guide/evm/using.md:27]: missing the new `block-keep-amount`

File: packages/vm/core/blocklog/external.go [hunk 3]
```diff
@@ -115,10 +112,32 @@ func GetEventsByBlockIndex(partition kv.KVStoreReader, blockIndex uint32, totalR
 	return ret
 }
 
-func GetBlockInfo(partition kv.KVStoreReader, blockIndex uint32) (*BlockInfo, error) {
-	ret, err := BlockInfoFromBytes(getBlockInfoBytes(partition, blockIndex))
+func PruneEventsByBlockIndex(partition kv.KVStore, blockIndex uint32, totalRequests uint16) []string {
```
Comment (jorgemmsilva) [inline packages/vm/core/blocklog/external.go]: should this be an "external" func? it's not used outside `blocklog`, right? probably could  be "not exported" and moved to `internal.go`

File: packages/vm/core/blocklog/internal.go [hunk 4]
```diff
@@ -20,6 +20,31 @@ func SaveNextBlockInfo(partition kv.KVStore, blockInfo *BlockInfo) {
 	registry.Push(blockInfo.Bytes())
 }
 
+func Prune(partition kv.KVStore, latestBlockIndex uint32, blockKeepAmount int32) {
```
Comment (jorgemmsilva) [inline packages/vm/core/blocklog/internal.go]: this one should probably be in `external.go`

File: packages/vm/core/blocklog/internal.go [hunk 5]
```diff
@@ -20,6 +20,31 @@ func SaveNextBlockInfo(partition kv.KVStore, blockInfo *BlockInfo) {
...
+	if blockKeepAmount <= 0 {
+		// keep all blocks
+		return
+	}
+	if latestBlockIndex < uint32(blockKeepAmount) {
+		return
+	}
+	toDelete := latestBlockIndex - uint32(blockKeepAmount)
+	// assume that all blocks prior to `toDelete` have been already deleted, so
+	// we only need to delete this one.
```
Comment (jorgemmsilva) [inline packages/vm/core/blocklog/internal.go]: we should maybe enfoce (prune those blocks) this when the "block to keep" setting is changed, so to not leave "zombie blocks"

**Review by jorgemmsilva — APPROVED** [review]

File: packages/vm/core/blocklog/internal.go (hunk 5 above)
Comment (dessaya) [reply packages/vm/core/blocklog/internal.go]: But how would that work? What would happen if blockKeepAmount was 10000 and it's changed to 10? Deleting 9990 blocks in a loop is a no-go IMO.

Comment (jorgemmsilva) [reply packages/vm/core/blocklog/internal.go]: >Deleting 9990 blocks in a loop is a no-go IMO.

why? :sweat_smile: too slow?

Comment (dessaya) [reply packages/vm/core/blocklog/internal.go]: Because of the size of the block. Isn't there a limit for the list of mutations in the block?

Comment (jorgemmsilva) [reply packages/vm/core/blocklog/internal.go]: there isn't afaik, but you're right. That would probably result in gigantic blocks, not a good thing.

**Pull request merged**